File: backend/relatorios.py
from backend.database import conectar


def relatorio_estoque():
    conn = conectar()

    produtos = conn.execute(
        """
        SELECT *
        FROM produtos
        WHERE ativo = 1
        ORDER BY nome
        """
    ).fetchall()

    conn.close()

    lista = []

    for produto in produtos:
        lista.append({
            "id": produto["id"],
            "nome": produto["nome"],
            "quantidade": produto["quantidade"],
            "preco": produto["preco"],
            "valor_total": round(produto["quantidade"] * produto["preco"], 2),
            "estoque_minimo": produto["estoque_minimo"],
            "abaixo_do_minimo": produto["quantidade"] <= produto["estoque_minimo"],
        })

    return lista


def relatorio_movimentacoes():
    conn = conectar()

    movimentacoes = conn.execute(
        """
        SELECT *
        FROM movimentacoes
        ORDER BY data_hora DESC
        """
    ).fetchall()

    conn.close()

    lista = []

    for m in movimentacoes:
        lista.append({
            "id": m["id"],
            "produto_id": m["produto_id"],
            "tipo": m["tipo"],
            "quantidade": m["quantidade"],
            "observacao": m["observacao"],
            "data_hora": m["data_hora"],
            "saldo_anterior": m["saldo_anterior"],
            "saldo_atual": m["saldo_atual"],
        })

    return lista

File: frontend/js/relatorios.js
// =========================
// Inicializacao
// =========================

document.addEventListener("DOMContentLoaded", async () => {

    await carregarRelatorioEstoque();

    await carregarRelatorioMovimentacoes();

});

// =========================
// Relatorio de estoque
// =========================

async function carregarRelatorioEstoque() {

    const tbody = document.getElementById("relatorioEstoque");

    tbody.innerHTML = "";

    try {

        const estoque = await listarRelatorioEstoque();

        if (estoque.length === 0) {

            tbody.innerHTML = `
                <tr>
                    <td colspan="5" style="text-align:center;">
                        Nenhum produto cadastrado.
                    </td>
                </tr>
            `;

            return;

        }

        estoque.forEach(produto => {

            tbody.innerHTML += `
                <tr>
                    <td>${produto.nome}</td>
                    <td>${produto.quantidade}</td>
                    <td>R$ ${Number(produto.preco).toFixed(2)}</td>
                    <td>R$ ${Number(produto.valor_total).toFixed(2)}</td>
                    <td class="${produto.abaixo_do_minimo ? "estoque-baixo" : ""}">
                        ${produto.abaixo_do_minimo ? "Estoque baixo" : "Normal"}
                    </td>
                </tr>
            `;

        });

    } catch (erro) {

        console.error(erro);

    }

}

// =========================
// Relatorio de movimentacoes
// =========================

async function carregarRelatorioMovimentacoes() {

    const tbody = document.getElementById("relatorioMovimentacoes");

    tbody.innerHTML = "";

    try {

        const produtos = await listarProdutos();
        const movimentacoes = await listarRelatorioMovimentacoes();

        const mapaProdutos = {};

        produtos.forEach(produto => {

            mapaProdutos[produto.id] = produto.nome;

        });

        if (movimentacoes.length === 0) {

            tbody.innerHTML = `
                <tr>
                    <td colspan="7" style="text-align:center;">
                        Nenhuma movimentacao registrada.
                    </td>
                </tr>
            `;

            return;

        }

        movimentacoes.forEach(movimentacao => {

            tbody.innerHTML += `
                <tr>
                    <td>${mapaProdutos[movimentacao.produto_id] || "Produto nao encontrado"}</td>
                    <td>${movimentacao.tipo}</td>
                    <td>${movimentacao.quantidade}</td>
                    <td>${movimentacao.saldo_anterior}</td>
                    <td>${movimentacao.saldo_atual}</td>
                    <td>${movimentacao.observacao || "-"}</td>
                    <td>${movimentacao.data_hora}</td>
                </tr>
            `;

        });

    } catch (erro) {

        console.error(erro);

    }

}
